test/controllers/UkimsNumberControllerSpec.scala

```diff
@@ -0,0 +1,97 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers
+
+import base.SpecBase
+import controllers.actions.FakeAuthoriseAction
+import forms.UkimsNumberFormProvider
+import org.scalatest.matchers.should.Matchers.convertToAnyShouldWrapper
+import play.api.test.FakeRequest
+import play.api.test.Helpers._
+import views.html.UkimsNumberView
+
+class UkimsNumberControllerSpec extends SpecBase {
+
+  private val formProvider = new UkimsNumberFormProvider()
+
+  private val fieldName = "ukimsNumber"
+
+  private val ukimsNumberView = app.injector.instanceOf[UkimsNumberView]
+
+  private val ukimsNumberController = new UkimsNumberController(
+    stubMessagesControllerComponents(),
+    new FakeAuthoriseAction(defaultBodyParser),
+    ukimsNumberView,
+    formProvider
+  )
+
+  "Ukims Number Controller" - {
+
+    "must return OK and the correct view for a GET" in {
+
+      val result = ukimsNumberController.onPageLoad(fakeRequest)
+
+      status(result) mustEqual OK
+
+      contentAsString(result) mustEqual ukimsNumberView(formProvider())(fakeRequest, stubMessages()).toString
+
+    }
+
+    "must redirect on Submit when user enters correct Ukims number" in {
+
+      val validUkimsNumber = "XI47699357400020231115081800"
+
+      val fakeRequestWithData = FakeRequest().withFormUrlEncodedBody(fieldName -> validUkimsNumber)
+
+      val result = ukimsNumberController.onSubmit(fakeRequestWithData)
+
+      status(result) mustEqual SEE_OTHER
+
+      // TO DO : Redirects needs to be change to actual when available
+      redirectLocation(result) shouldBe Some(routes.DummyController.onPageLoad.url)
+
+    }
+
+    "must send bad request on Submit when user leave the field blank" in {
+
+      val formWithErrors = formProvider().bind(Map.empty[String, String])
+
+      val result = ukimsNumberController.onSubmit(fakeRequest)
+
+      status(result) mustEqual BAD_REQUEST
+
+      contentAsString(result) mustEqual ukimsNumberView(formWithErrors)(fakeRequest, stubMessages()).toString
+
+    }
+
+    "must send bad request on Submit when user enters invalid ukims number" in {
+
+      val invalidUkimsNumber = "XIAA476993574000202311"
+
+      val formWithErrors = formProvider().bind(Map(fieldName -> invalidUkimsNumber))
+
+      val fakeRequestWithData = FakeRequest().withFormUrlEncodedBody(fieldName -> invalidUkimsNumber)
+
+      val result = ukimsNumberController.onSubmit(fakeRequestWithData)
+
+      status(result) mustEqual BAD_REQUEST
+
+      contentAsString(result) mustEqual ukimsNumberView(formWithErrors)(fakeRequest, stubMessages()).toString
```

Comment:
it would be good for this test to verify that it returns the correct error message for this error type

Reply:
You could do this in the integration tests like this:

`html(result) should include("Enter your UKIMS number in the correct format should be displayed")`

in the unit test you could use mustContain <errorMessage> instead of mustEqual